sinatra-example.rb

```diff
@@ -17,4 +19,73 @@ class SinatraExample < Sinatra::Base
     "slept for #{milliseconds} milliseconds"
   end
 
-end
\ No newline at end of file
+  get '/mem/alloc/:size_mb/?:leak?/?:once?' do
+    if params[:once] and params[:once]=1 and $leak_buffer
+      puts "Already has memory allocated, skipping"
+      return "Already has memory allocated, skipping"
+    end
+
+    size_mb = params[:size_mb].to_i
+    total_bytes = size_mb*1024**2
+    do_leak = (params[:leak] and params[:leak] == "1")
+
+    # Metrics before allowcate
+    rss_bytes_before = OS.rss_bytes
+    free_mem_before = (Vmstat.memory.free * Vmstat.memory.pagesize) / 1024
+
+    puts "Allocating: #{size_mb} MB. Leaking memory: #{if do_leak; "Yes" else  "No" end}, system free: #{free_mem_before/1024} MiB, Process rss: #{rss_bytes_before/1024} MiB"
+
+    # Allocate the memory chunk
+    chunk_of_mem = 'a'*(total_bytes-1)
+    if do_leak
+        $leak_buffer = [] unless $leak_buffer
+        $leak_buffer << chunk_of_mem
+    end
+
+    sleep(1.0/10.0) # To be sure that Vmstat will catch the memory change
+
+    # Metrics after allocate
+    rss_bytes_after = OS.rss_bytes
+    free_mem_after = (Vmstat.memory.free * Vmstat.memory.pagesize) / 1024
+
+    puts "Allocated: #{size_mb} MB. Leaking memory: #{if do_leak; "Yes" else  "No" end}, system free: #{free_mem_after/1024} MiB, Process rss: #{rss_bytes_after/1024} MiB"
+
+    """Allocated #{size_mb} MB = #{total_bytes} bytes.
+Leaking memory: #{if do_leak; "Yes" else  "No" end}
+Free system memory before: #{free_mem_before} kiB aka #{free_mem_before/1024} MiB
+Free system memory after: #{free_mem_after} kiB aka #{free_mem_after/1024} MiB
+Process memory before: #{rss_bytes_before} kiB aka #{rss_bytes_before/1024} MiB
+Process memory after: #{rss_bytes_after} kiB aka #{rss_bytes_after/1024} MiB
+"""
```

Comment:
That is because it returns a plain text and the browser tries to just print it as HTML

I did not want to spend time on this, it is just a test app that we run locally.